Remove dependencies on Tree-sitter grammars

Xophmeister opened the pull request:
# Remove dependencies on Tree-sitter grammars

<!----------------------------------------------------------------------
If this PR is related to, or resolves an issue, please reference it
here.

- For issues that are fixed by this PR, use GitHub's automatic closing
  feature by writing "Resolves #XXX", for example.

- If multiple issues are implicated, please list them out, with an
  appropriate byline for each issue, with one issue per line.

- You may also use this section to reference other PRs and discussion
  items, following the same pattern outlined here for issues.

Below is an example; please update as appropriate. If no issue, etc. is
implicated in this PR, please remove this section.
----------------------------------------------------------------------->
Resolves #1022
Resolves #1023

## Description

Remove the development dependencies on `tree-sitter-json` and `tree-sitter-nickel`. Instead, use the dynamic loading from `topiary-config` in the relevant tests.

## Checklist

<!----------------------------------------------------------------------
See MAINTAINERS.md for more details.
This is particularly important if this PR is preparing a release.
----------------------------------------------------------------------->

Checklist before merging, wherever relevant:

- [x] `CHANGELOG.md` updated
- [x] Documentation (The Topiary Book, `README.md`, etc.) up-to-date
<!----------------------------------------------------------------------
If the PR solves a formatting issue for a supported language,
or generally improves formatting, please make sure to include a
regression test showcasing the fix/improvement to:
`topiary-cli/tests/samples/{input,expected}/mylanguage.lang`
----------------------------------------------------------------------->
- [ ] ~Updated regression tests~

Xophmeister commented:
These changes work outside of Nix, but the CI complains inside (e.g., see [here](https://github.com/tweag/topiary/actions/runs/15687381415/job/44193857267?pr=1032)). My best guess is: Now that `topiary-core` depends on `topiary-config` and `topiary-queries`, for testing, I _think_ that the derivation under test is simply missing the query files. That is, the "permission denied" error is misleading and should be, rather, "file not found".

Does that sound plausible, @Niols?

Adding the JSON and Nickel query files to the `topiary-core` package doesn't sound like the correct solution, as they're only needed for testing. What would be the right approach?

Xophmeister commented:
> > Adding the JSON and Nickel query files to the `topiary-core` package doesn't sound like the correct solution, as they're only needed for testing. What would be the right approach?
> 
> We could disable tests in the Nix package, but that's a bit sad. We could also make `topiary-core` 's Nix package depend on `topiary-queries` (doesn't it sort of already, if there is such a dependency in Rust as well?) and point (don't know how yet) the tests at the right path in the store?

We currently rely on the tests to run in the CI, so disabling them is a bit of a non-starter. `topiary-queries` is a dev-dependency of `topiary-core`, so perhaps there's scope for two Nix packages: one with tests -- which we invoke in the CI -- and one without, which downstream users can benefit from... I don't know how "idiomatic" that would be :shrug: 

@yannham and @jneem: Have you any experience in this, from Nickel or elsewhere?

Niols commented:
Nix already has this sort of distinction between build dependencies and runtime dependencies. Downstream users would presumably just download the pre-built binaries from a Nix cache, which will also require runtime dependencies, but not build dependencies.

mkatychev commented:
Hi @Xophmeister, is this PR still needed to be reviewed and relevant?

Xophmeister commented:
Thanks for looking through these old PRs, @mkatychev :pray:

It's definitely still relevant, but it's so old now that there's some serious divergence and, IIRC, I was never able to address the Nix code. I'll need to look into it a bit more closely, but I suspect I might be able to get this over the line, if time permits and the rebase doesn't drive me crazy :sweat_smile:

I'll add it to my todo list :+1:

Xophmeister committed:
Remove Tree-sitter grammar crates as test dependencies

Replace the `topiary-core` test and benchmark dependencies on the
`tree-sitter-json` and `tree-sitter-nickel` grammar crates with dynamic
grammar loading via `topiary-config`, mirroring how the OCaml grammars
are already loaded in the tests and the injections benchmark.

- Workspace: drop the `tree-sitter-json` and `tree-sitter-nickel` deps.
- topiary-core: drop the two grammar dev-deps; enable the `json` and
  `nickel` features on `topiary-config`/`topiary-queries` so the
  grammars can be loaded dynamically and their queries embedded.
- lib.rs: the `formatter` doctest and the
  `parsing_error_fails_formatting` / `tolerate_parsing_errors` tests now
  obtain the JSON grammar from `topiary_config::Configuration`.
- benchmark.rs: the Nickel `format` benchmark now loads its grammar the
  same way.

Resolves #1022
Resolves #1023

Xophmeister commented:
I've rebased this on current `main` and it looks like the Nix refactoring done in #1026 has solved the problem that was blocking this PR for free! Thanks, @Niols :pray:

This is now ready for review :tada:

Xophmeister commented:
GitHub won't let me change reviewers, probably because the `topiary-core-team` was a Tweag team and the org migration doesn't account for that... So pinging @topiary/core @topiary/honorary

mkatychev reviewed (approved):
What a fantastic change, makes everything more consistent! Thank you for your effort!

mkatychev commented on Cargo.toml:
```diff
@@ -62,9 +62,7 @@ thiserror = "2.0.16"
 toml = "1.0"
-tree-sitter-json = "0.24"
 tree-sitter-language = "0.1"
-tree-sitter-nickel = "0.5"
```
💪
